tests/test_serialisation.py

```diff
@@ -79,6 +81,61 @@ def test_leaf_serialisation_file(getkey, tmp_path):
     assert tree_loaded_obj is like_obj
 
 
+def test_helpful_errors(getkey, tmp_path):
+    # Test that we get a helpful error message when the loading itself fails
+    tree = jnp.array(1), {}
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    bad_like_tree = jnp.array(2), {
+        "a": jnp.array(2),
+        "b": jnp.array(2),
+    }
+    with pytest.raises(
+        EOFError,
+        match=r"Error at leaf with path \(SequenceKey\(idx=1\), DictKey\(key='a'\)\)",
+    ):
+        _ = eqx.tree_deserialise_leaves(tmp_path, bad_like_tree)
+
+    # Test that we get a helpful error message when the types don't match
+    tree = jnp.array(1), {
+        "a": jnp.array(1),
+        "b": jnp.array(1),
+    }
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    bad_like_tree = jnp.array(2), {
+        "a": jnp.array(2, dtype=jnp.float32),
+        "b": jnp.array(2),
+    }
+    with pytest.raises(
+        RuntimeError,
+        match=r"Deserialised leaf at path \(SequenceKey\(idx=1\), DictKey\(key='a'\)\)",
+    ):
+        _ = eqx.tree_deserialise_leaves(tmp_path, bad_like_tree)
+
+
+def test_generic_dtype_serialisation(getkey, tmp_path):
+    # Ensure we can round trip when we start with an array
+    jax_array = jnp.array(bfloat16(1))
+    eqx.tree_serialise_leaves(tmp_path, jax_array)
+    like_jax_array = jnp.array(bfloat16(2))
+    loaded_jax_array = eqx.tree_deserialise_leaves(tmp_path, like_jax_array)
+    assert eqx.tree_equal(jax_array, loaded_jax_array)
+
+    tree = jnp.array(1), bfloat16(1), np.float32(1), jnp.array(1)
+    like_tree = jnp.array(2), bfloat16(2), np.float32(2), jnp.array(2)
+
+    # Ensure we can round trip when we start with a scalar
+    eqx.tree_serialise_leaves(tmp_path, tree)
+    loaded_tree = eqx.tree_deserialise_leaves(tmp_path, like_tree)
+    # Can't use `tree_equal` because `is_array` treats `bfloat16` as an array and
+    # ends up trying to call `all()` on a `bool`
+    assert loaded_tree == tree
```

Comment:
FWIW you may prefer to use `from helpers import shaped_allclose; assert shaped_allclose(loaded_tree, tree)` here.

This helper does strict everything-must-match: shape, dtype, array type, etc. In contrast `__eq__` is fairly permissive, e.g. it happily promotes arraylikes.

(Side note, the name `shaped_allclose` is a bit odd -- I've been meaning to rename it to `tree_identical` or something.)